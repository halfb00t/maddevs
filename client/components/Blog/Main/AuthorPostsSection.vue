<template>
  <div class="author-posts">
    <div class="container">
      <div
        class="author-posts__featured-post"
        :class="[authorPosts.length === 1 ? 'author-posts__featured-post--mb-0' : '']"
      >
        <FeaturedPost
          v-if="authorPostsLoaded && authorPosts && authorPosts.length"
          :post="authorPosts[0]"
          :author="blogAuthor"
          theme="light"
          :disable-author-link="true"
        />
        <SkeletonFeaturedPost
          v-else
          theme="light"
        />
      </div>
      <div class="author-posts__list">
        <template v-if="authorPostsLoaded">
          <PostCard
            v-for="post in authorPostsToShow"
            :key="post.id"
            :post="post"
            :author="blogAuthor"
            :disable-author-link="true"
            class="author-posts__list-item"
          />
        </template>
        <template v-else>
          <section
            v-for="i in 6"
            :key="i"
            class="author-posts__list-item"
          >
            <SkeletonBlogWidget />
          </section>
        </template>
      </div>
      <div
        v-if="totalPages > authorPostsPage"
        class="author-posts__load-more"
      >
        <LoadMoreButton
          @click="getMoreAuthorPosts"
        />
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import FeaturedPost from '@/components/Blog/shared/FeaturedPost'
import SkeletonFeaturedPost from '@/components/Blog/skeletons/SkeletonFeaturedPost'
import SkeletonBlogWidget from '@/components/Blog/skeletons/SkeletonBlogWidget'
import PostCard from '@/components/Blog/shared/PostCard'
import LoadMoreButton from '@/components/Blog/shared/LoadMoreButton'
import initializeLazyLoad from '@/helpers/lazyLoad'

export default {
  name: 'AuthorPostsSection',
  components: {
    FeaturedPost,
    SkeletonFeaturedPost,
    SkeletonBlogWidget,
    PostCard,
    LoadMoreButton,
  },

  data() {
    return {
      pageSize: 7,
    }
  },

  computed: {
    ...mapGetters(['blogAuthor', 'authorPosts', 'authorPostsLoaded', 'authorPostsPage']),

    authorPostsToShow() {
      if (this.authorPosts && !this.authorPosts.length) return []
      return this.authorPosts.slice(0, this.pageSize * this.authorPostsPage)
    },

    totalPages() {
      return Math.ceil(this.authorPosts.length / this.pageSize)
    },
  },

  updated() {
    this.$nextTick(() => initializeLazyLoad())
  },

  methods: {
    ...mapActions(['getMoreAuthorPosts']),
  },
}
</script>

<style lang="scss" scoped>
  .author-posts {
    background-color: $bgcolor--white-primary;
    padding: 60px 0 71px;

    &__featured-post {
      margin-bottom: 137px;
      &--mb-0 {
        margin-bottom: 0;
      }
    }

    ::v-deep .post-card {
      display: flex;
      flex-direction: column;
      width: 33.3333%;

      @media screen and (max-width: 991px) {
        width: 100%;
      }
      &__info {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
      }
      &__info-text{
        margin-bottom: 16px;
      }
    }

    &__list {
      display: flex;
      flex-flow: row wrap;
      margin: 0 -10px;
      row-gap: 103px;
    }

    &__list-item {
      box-sizing: border-box;
      width: 33.3333%;
      padding: 0 10px;
      display: flex;
      flex-direction: column;
      height: auto;
      ::v-deep .post-card__meta {
        margin: auto 0 16px;
      }
      &:first-of-type {
        display: none;
      }
    }

    &__load-more {
      margin-top: 75px;
      ::v-deep .load-more-button {
        width: 100%;
      }
    }

    @media only screen and (max-width: 991px) {
      padding: 34px 0 60px;

      &__featured-post {
        display: none;
        margin-bottom: 56px;
      }

      &__list {
        row-gap: 56px;
      }

      &__list-item {
        width: 100%;
        &:first-of-type {
          display: block;
        }
      }

      &__load-more {
      margin-top: 56px;
    }
    }
  }
</style>
