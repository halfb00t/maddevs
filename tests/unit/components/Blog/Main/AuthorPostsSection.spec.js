import { render, screen, fireEvent } from '@testing-library/vue'
import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import AuthorPostsSection from '@/components/Blog/Main/AuthorPostsSection'
import allPosts from '../../../__mocks__/allPosts'
import initializeLazyLoad from '@/helpers/lazyLoad'
// eslint-disable-next-line import/order
import lazyLoad from 'nuxt-lazy-load/lib/module'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('lazy-load', lazyLoad)
jest.mock('nuxt-lazy-load/lib/module')

const mocks = {
  $prismic: {
    asText: () => 'img.png',
  },
}

const store = {
  getters: {
    blogAuthor: () => ({
      name: 'Maddevs',
      postition: 'position',
      image: {
        alt: 'alt',
        url: 'someurl',
      },
    }),
    authorPosts: () => allPosts,
    authorPostsLoaded: () => true,
    authorPostsPage: () => 1,
  },
}

const stubs = ['PrismicImage', 'NuxtLink', 'FeaturedPost', 'PostCard']

jest.mock('@/helpers/lazyLoad', () => ({
  initializeLazyLoad: jest.fn(),
}))

describe('AuthorPostsSection component', () => {
  it('should render correctly', async () => {
    const initLazyLoadMock = jest.spyOn(initializeLazyLoad, 'initializeLazyLoad').mockImplementation(() => {})
    const nextTick = jest.fn(() => initLazyLoadMock())
    mocks.$nextTick = nextTick
    const { container, getAllByTestId } = render(AuthorPostsSection, {
      stubs,
      mocks,
      store,
      localVue,
    })

    const button = screen.getByTestId('test-load-more-button')
    await fireEvent.click(button)
    expect(getAllByTestId('author-posts__list-item')).toHaveLength(14)
    expect(container).toMatchSnapshot()
  })

  it('should render correctly with empty array authorPosts', async () => {
    store.getters.authorPosts = () => []
    const initLazyLoadMock = jest.spyOn(initializeLazyLoad, 'initializeLazyLoad').mockImplementation(() => {})
    const nextTick = jest.fn(() => initLazyLoadMock())
    mocks.$nextTick = nextTick

    const wrapper = mount(AuthorPostsSection, {
      mocks,
      store,
      stubs,
      localVue,
    })

    expect(wrapper.exists()).toBe(true)
    await wrapper.vm.$forceUpdate()
    await wrapper.vm.$nextTick(() => {
      expect(initLazyLoadMock).toHaveBeenCalledTimes(1)
    })
  })
})
