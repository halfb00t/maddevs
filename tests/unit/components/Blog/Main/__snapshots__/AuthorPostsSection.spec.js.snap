// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorPostsSection component should render correctly 1`] = `
<div>
  <div
    class="author-posts"
  >
    <div
      class="container"
    >
      <div
        class="author-posts__featured-post"
      >
        <featuredpost-stub
          author="[object Object]"
          disable-author-link="true"
          post="[object Object]"
          theme="light"
          titletag="h2"
        />
      </div>
       
      <div
        class="author-posts__list"
      >
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
        <section
          class="author-posts__list-item"
          post="[object Object]"
        >
          <postcard-stub
            author="[object Object]"
            disableauthorlink="true"
            post="[object Object]"
            theme="light"
          />
        </section>
      </div>
       
      <div
        class="author-posts__load-more"
      >
        <button
          class="load-more-button"
          data-testid="test-load-more-button"
        >
          
  See more

        </button>
      </div>
    </div>
  </div>
</div>
`;
